go/internal/services/keys/verifier.go

```diff
@@ -0,0 +1,135 @@
+	Cost       int64
+	Name       string
+	Duration   time.Duration
+	Limit      int64
+	AutoApply  bool
+	Identifier string                       // The identifier to use for this rate limit
+	Response   *ratelimit.RatelimitResponse // nil until rate limit is checked
+}
+
+// KeyVerifier represents a key that has been loaded from the database and is ready for verification.
+// It contains all the necessary information and services to perform various validation checks.
+type KeyVerifier struct {
+	Key                   db.FindKeyForVerificationRow                  // The key data from the database
+	ratelimitConfigs      map[string]db.KeyFindForVerificationRatelimit // Rate limits configured for this key (name -> config)
+	Roles                 []string                                      // RBAC roles assigned to this key
+	Permissions           []string                                      // RBAC permissions assigned to this key
+	Valid                 bool                                          // Whether the key passed all validation checks
+	Status                KeyStatus                                     // The current validation status
+	AuthorizedWorkspaceID string                                        // The workspace ID this key is authorized for
+	RatelimitResults      map[string]RatelimitConfigAndResult           // Combined config and results for rate limits (name -> config+result)
+	isRootKey             bool                                          // Whether this is a root key (special handling)
+	session               *zen.Session                                  // The current request session
+	rateLimiter           ratelimit.Service                             // Rate limiting service
+	usageLimiter          usagelimiter.Service                          // Usage limiting service
+	rBAC                  *rbac.RBAC                                    // Role-based access control service
+	clickhouse            clickhouse.ClickHouse                         // Clickhouse for telemetry
+	logger                logging.Logger                                // Logger for verification operations
+	message               string                                        // Internal message for validation failures
+}
+
+// GetRatelimitConfigs returns the rate limit configurations (for internal use)
+func (k *KeyVerifier) GetRatelimitConfigs() map[string]db.KeyFindForVerificationRatelimit {
+	return k.ratelimitConfigs
+}
+
+// Verify performs key verification with the given options.
+// For root keys: returns fault errors for validation failures.
+// For normal keys: returns error only for system problems, check k.Valid and k.Status for validation results.
+func (k *KeyVerifier) Verify(ctx context.Context, opts ...VerifyOption) error {
+	// Skip verification if key is already invalid
+	if !k.Valid {
+		// For root keys, auto-return validation failures as fault errors
+		if k.isRootKey {
+			return k.ToFault()
+		}
+		return nil
```

Comment:
I don't understand this. Say I'm verifying a regular key, and do `err = k.Verify(ctx)` Now my err is actually nil, which to me sounds like we should proceed with the actual api handler.

How is returning nil the correct move here?

Reply:
When we  verify a normal key, we want errors where something genuinely went wrong, e.g rate limiting, or updating usage so that we can return a 5xx response.

In any other case e.g key not valid / no perms we use the status / soon to be removed valid field to reflect that for example: https://github.com/unkeyed/unkey/pull/3474/files#diff-0af8b838792fcf37d741f10a6ce245344a7d175b63169a305337fbea1fc225ceR136